### DISCUSSION TITLE
Configuration loading failing with 5.4.2 and Docker

### DISCUSSION THREAD

**Discussion by ghost** [mikro-orm/mikro-orm#3506] [Q&A]
**Describe the bug**
I am working on a Next.js app using mikro-orm for database access and schema management. All works well with Next.js' dev server and running the production code natively on my Linux workstation. Trying to run the same code fails during `MikroOrm.init()` with an error from `acorn-walk`:

**Stack trace**
```
TypeError: baseVisitor[type] is not a function
    at c (/app/node_modules/acorn-walk/dist/walk.js:29:24)
    at Object.base.ClassBody (/app/node_modules/acorn-walk/dist/walk.js:441:7)
    at c (/app/node_modules/acorn-walk/dist/walk.js:29:24)
    at Object.base.Class (/app/node_modules/acorn-walk/dist/walk.js:434:5)
    at c (/app/node_modules/acorn-walk/dist/walk.js:29:24)
    at Object.base.ClassDeclaration.base.ClassExpression (/app/node_modules/acorn-walk/dist/walk.js:430:82)
    at c (/app/node_modules/acorn-walk/dist/walk.js:29:24)
    at Object.skipThrough (/app/node_modules/acorn-walk/dist/walk.js:186:39)
    at c (/app/node_modules/acorn-walk/dist/walk.js:29:24)
    at Object.base.Program.base.BlockStatement (/app/node_modules/acorn-walk/dist/walk.js:198:7)
```

This same code was working before we upgraded to 5.4.0. We upgraded again to 5.4.2 to pick up some fixes for Typescript and the CLI so are kind of past the point of a downgrade being an option

**To Reproduce**
1. Entities and config are written in Typescript but compiled and copied into a Docker image
2. The entry point for the image calls the server.js module (we use a custom server with Next.js) and that code handles calling `MikroOrm.init()`
3. Try to start the image and see the error during configuration loading

**Expected behavior**

**Additional context**
- I copied the same files in the docker container to a fresh directory--our compiled JS, Next.js's built .next/ folder, node_modules/, and our package.json into a fresh folder on my Linux workstation
- I ran the script entry we use in our Docker image and initialization succeeded as expected so I am worried it is something to do with acorn and Docker's file system
- I already had another Docker issue: I could not get cache generation to work in our multistage Docker build so had to generate outside the container and copy in. I adjusted that to make sure that the cache files are `*.js.json` instead of `*.ts.json` like when the CLI uses the TS version of the config but that didn't seem to make a difference

**Versions**

| Dependency | Version |
| - | - |
| node | 16 |
| typescript | 4.7 |
| mikro-orm | 5.4.2 |
| your-driver | PostgreSQL |

**Comment by B4nan** [comment]
I'd say you should report this to acorn, together with a reproduction of the code that breaks it (just add a console log before `acorn` gets called to see, acorn is a parsing library so this is just about the code you pass into it), the stack trace is purely about that. Yes, we use it for parsing JS. Yes, we used different library in 5.4 (that is no longer maintained and has outdated deps). 

Just checked the acorn issues and found this one, maybe the workaround helps? https://github.com/acornjs/acorn/issues/829

The code that uses acorn is here: https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/Utils.ts#L368-L400

Can't really help with such issues without complete reproduction, there is too many moving parts, but the main issue about acorn is something that needs to be fixed upstream.

edit: if this won't help, as a last resort we can ignore such parsing errors, print a warning and continue. If you use constructors in your entities, `em.create()` could break a bit, but internally it shouldnt matter.

**Reply by ghost** [reply]
I will create a complete reproduction for you since I don't feel equipped to report this further upstream--our codebase has no direct dependency on acorn and I honestly don't have the time to understand Mikro-ORM's usage enough to work with the acorn maintainers towards a full resolution. I understand why you are suggesting it but don't think it is a realistic option.

I also found that issue but lacking a deeper understanding of how and why Mikro-ORM uses acorn, it wasn't clear to me what, if anything, I could do in my own code to try to workaround the issue. Are you suggesting I fork Mikro-ORM to add the suggested code in `Utils.ts`? Or is there some extension point or config option I am missing that would allow some form of the workaround to be applied in our own code?

Maybe to back up, what in `mikro-orm.config.ts`/`mikro-orm.config.js` affects the input to acorn? This is our config:

```typescript
import { TsMorphMetadataProvider } from "@mikro-orm/reflection";
import { loadEnvConfig } from "@next/env";

import { ENTITIES } from "./api/common/entities";

loadEnvConfig(process.cwd());

export default {
  type: "postgresql",
  entities: ENTITIES,
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  clientUrl: process.env.DATABASE_URI!,
  debug: process.env.DEBUG_SQL !== undefined,
  metadataProvider: TsMorphMetadataProvider,
  cache: {
    options: {
      cacheDir: "./.mikro-orm.cache/",
    },
  },
  migrations: {
    path: undefined,
    pathTs: "./migrations",
    disableForeignKeys: false,
  },
} as const;
```

We aren't using directory based discovery but, as the repo I'll create will show, for prod, using Docker, we do compile the config and use that version along with the declaration files for the entities. Odd that I can only reproduce this in a Docker container, so far.

**Reply by B4nan** [reply]
> I will create a complete reproduction for you since I don't feel equipped to report this further upstream--our codebase has no direct dependency on acorn and I honestly don't have the time to understand Mikro-ORM's usage enough to work with the acorn maintainers towards a full resolution. I understand why you are suggesting it but don't think it is a realistic option.

I asked you put a single `console.log()` statement into the ORM code to simplify the debugging, I am sure you can do that, it will take a minute, unlike building a complete reproduction :] I gave you the exact location where to put it - I need to see the exact input that gets passed to `acorn`. 

If you can give me that, its enough to reproduce your problem for me, I dont need anything else.

> Maybe to back up, what in mikro-orm.config.ts/mikro-orm.config.js affects the input to acorn? This is our config:

Nothing, you cant control this from configuration. This is about detecting constructor parameters in your entity definitions. 

As mentioned above, worst case we can ignore errors during parsing and act as if there are no constructor parameters. Also, technically we can extract the information we need via the `TsMorphMetadataProvider` (but its currently not implemented that way), so even if we would have to silence errors from acorn, we might have a way to support constructor detection for your use case.

**Reply by ghost** [reply]
@B4nan Sorry, I misunderstood your suggestion. I will  do as you advise and get you the output.

Thank you for the additional explanation, it wasn't entirely clear from your first comment how constructors intersected with this issue. It is clearer to me now and I will review our entities to check for constructors. I don't think we added any, yet, our project is still in early stages but it is possible someone else on the team did.

Thank you for your patience and sorry I let my frustration get the better of me. I'll get you the additional info ASAP.

**Reply by B4nan** [reply]
Regarding the acorn issue and workaround stated there, all I was asking for is to try adjusting [the code](https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/Utils.ts#L368-L400) like this:

```diff
walk(parsed, {
  MethodDefinition: (node: any) => checkNode(node, methodName),
  FunctionDeclaration: (node: any) => checkNode(node, methodName),
+ JSXElement: () => {},
});
```

> I don't think we added any

Ok that's quite possible that it fails even without the constructors being actually used. I was just trying to explain what that part of the codebase does and why/for what we use acorn.

**Reply by B4nan** [reply]
Btw if this is blocking you, I am sure you could monkey patch the ORM - if you dont use any constructors, its as easy as replacing the method to always return empty array. Something like this (be sure to run it before the ORM discovery):

```ts
import { Utils } from '@mikro-orm/core';

Utils.getParamNames = () => [];
```

**Reply by ghost** [reply]
@B4nan Thanks again for clarifying, this makes more and more sense to me now. Again, thanks for your patience and sorry for coming on strong. Aside from this and one other issue (which your team resolved pretty quickly), I have been thrilled with our experiences using Mikro-ORM. People rarely complain unless they care. ;)

While I am getting you the log output you asked for, I will try both of your suggestions. We aren't entirely blocked, our team can still develop locally without issue. This is breaking our lower environments in AWS, however. We will launch early November, so I have some time but the lack of shared dev and staging isn't great. I have had to do similar patches so appreciate the short term workaround while you look into this.

**Reply by ghost** [reply]
@B4nan Still working on adding the logs. Complicated a bit by node_modules containing the compiled JS to start though I figured out the equivalent compiled code based on the TS source. Now working to get the edited JS into our Docker image which ordinarily does a fresh `yarn install`.

I did confirm we have not added any constructors to our entities. We do however have a couple with onUpdate and onCreate hooks, in case that is relevant or intersects with the constructor analysis.

**Reply by B4nan** [reply]
I dont think you need to reproduce this from inside docker in your production environmnet, this will be probably the same if you run a production build locally and try to use that. The problem will be in running the compiled code instead of TS, its not about docker.

**Reply by ghost** [reply]
[bug.log](https://github.com/mikro-orm/mikro-orm/files/9575539/bug.log)

```javascript
    /**
     * Returns array of functions argument names. Uses `escaya` for source code analysis.
     */
    static getParamNames(func, methodName) {
        console.debug("Function", func.toString(), "methodName", methodName);
        const ret = [];
        const parsed = (0, acorn_loose_1.parse)(func.toString(), { ecmaVersion: 2022 });
        const checkNode = (node, methodName) => {
            if (methodName && !(node.key && node.key.name === methodName)) {
                return;
            }
            const params = node.value ? node.value.params : node.params;
            ret.push(...params.map((p) => {
                switch (p.type) {
                    case 'AssignmentPattern':
                        if (p.left.type === 'ObjectPattern') {
                            return exports.ObjectBindingPattern;
                        }
                        return p.left.name;
                    case 'RestElement':
                        return '...' + p.argument.name;
                    default:
                        return p.name;
                }
            }));
        };
        console.debug("Parsed", parsed);
        (0, acorn_walk_1.simple)(parsed, {
            MethodDefinition: (node) => checkNode(node, methodName),
            FunctionDeclaration: (node) => checkNode(node, methodName),
        });
        return ret;
    }
```
I think this may be related to doing `yarn install` within Docker, honestly. I had to change the Dockerfile to copy node_modules from the host system so my edits to add the console logging worked. The error doesn't occur when using the version from my host system, the server starts as expected. I am going to take a close look at the installed versions and any differences from the build inside of Docker versus natively.

**Reply by ghost** [reply]
@B4nan Yes except I can *only* reproduce the issue in Docker but your debug suggestion may have given me a key insight, per my previous comment. If I run the exact same code on my host system, also Linux, everything works as expected. I *thought* had had tried rebuilding the Docker image with caching disabled but will check again, in case that is causing a version mismatch

**Reply by ghost** [reply]
@B4nan I finally reproduced the problem outside of Docker and have a new theory. We use `clean-modules` because we are using a custom server so cannot rely solely on Next.js's file tracing and dependency management to keep our image smaller. If I follow the same steps as our Dockerfile, then I get the error on starting up. Let me try without `clean-modules` and go from there.

**Reply by ghost** [reply]
@B4nan I have confirmed, it is our use of `clean-modules` that is causing the different behavior in Docker, not anything in mikro-orm. Thanks again for your patience, especially for letting me think this through with you. I realize you must get a ton of these kinds of messages and issues from frustrated developers. Hopefully this discussion helps anyone else using `clean-modules` or similar as part of their production build process.

**Reply by B4nan** [reply]
Hmm but that logs dont contain your error from the opening post. This will help only if you provide logs before the exception.

Have you tried building a production bundle and running that locally?

**Reply by ghost** [reply]
I have now and realized the error is caused by our use of `clean-modules`. That tool is removing something that trips the error. That is why the log file attachment doesn't have the error, the code that produced that used my local node_modules, against which I had *not* run clean-modules. If I run clean-modules and try to start the app locally, the problem happens. I can try to get you a log with a version that has been "cleaned", if you are willing to help me troublehshoot further but it is OK if you want to close this discussion. This is clearly a problem of our own making, not strictly a bug.

**Reply by B4nan** [reply]
I can try to look at the `clean-modules` package, maybe I can reproduce easily. If you can provide the (broken) input that gets passed to `acorn`, it would speed up things for sure.

EIther way, this is definitely something to future-prove, by ignoring parsing errors if they happen in this part of app, as it does not have to mean the app wont work as expected. At least I will add the try/catch block to fail silently.

**Reply by ghost** [reply]
OK, let me try to get you the debug in the breaking case. I suspect you should be able to reproduce pretty easily, our call to the tool is simply `yarn clean-modules -y -i node_modules/`. We run that after `yarn install` and `yarn next build` but before starting the server. I can still try to pull together a small repo sanitized out of our code if you aren't able to reproduce easily. Since you are willing to keep looking into this, I want to help as well to make mikro-orm more resilient. Again, my original frustration was at least because in every other way, our experience using mikro-orm over the past six months has been pretty great

**Reply by B4nan** [reply]
I never used next.js, so that's the part I would have hard times reproducing, I dont have any existing project I could just try this on. 

But I can definitely try to run `clean-modules` on some of the example apps I have (regular backend ups, e.g. nestjs or express).

**Reply by ghost** [reply]
https://github.com/thomas-yaa/clean-modules-example

This does not yet reproduce the issue. Something else in our actual project must be involved.

- `yarn install --immutable`
- `yarn build`
- `yarn clean-modules -y -i node_modules/ -i .yarn/cache/`
- `yarn start`

In our actual project, that last step is where the error occurs. `src/entities.ts` is exactly the source from our project which I am mildly uncomfortable with but without the rest of the code is probably fine. I initially took out the contract and vehicle entities and the enums since I felt they were specific to our codebase and not likely to make a difference.

**Reply by ghost** [reply]
Some other tools we use in our actual project that maybe are part of the cause:
- ajv - we use to validate requests from our own front end, we have some larger schemas and use the library's ability to cache and share schemas
- mapped paths - our codebase is already large enough that having a few top level mappings makes imports easier to manage, I am less sure this has anything to do with it; I took great care from our custom server code to ensure none of the imports, direct or indirect, used those paths, making sure they used relative path imports instead

I can pretty quickly add sanitized config and code for either if you think they are likely to shed further light

**Reply by B4nan** [reply]
I'd suggest you try to provide the code that breaks your app from production if its hard to reproduce locally. You could just use [patch-package](https://github.com/ds300/patch-package), its very convenient, you modify the module locally, create a patch and ship it into your production.

You could add a try catch around the `acorn` call, and print the input only in case of failure.
